monai/inferers/merger.py

```diff
@@ -0,0 +1,186 @@
+
+        Raises:
+            NotImplementedError: When the subclass does not override this method.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def aggregate(self, values: torch.Tensor, location: Sequence[int]) -> None:
+        """
+        Aggregate values for merging.
+
+        Args:
+            values: a tensor of shape BCHW[D], representing the values of inference output
+            location: a tuple/list giving the top left location of the patch in the original image.
+
+        Raises:
+            NotImplementedError: When the subclass does not override this method.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def finalize(self) -> torch.Tensor:
+        """
+        Perform final operations for merging patches.
+
+        Raises:
+            NotImplementedError: When the subclass does not override this method.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def get_output(self) -> torch.Tensor:
+        """
+        Get the output merged tensor.
+
+        Raises:
+            NotImplementedError: When the subclass does not override this method.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class AvgMerger(Merger):
+    """Merge patches by taking average of the overlapping area
+
+    Args:
+        output_shape: optional fix size for the shape of the output tensor. Default to None.
+            If provided, it overwrites the `output_shape` in `initialize(output_shape)`.
+        device: the device for aggregator tensors and final results.
+        dtype: the dtype for aggregation and final result and .
+    """
+
+    def __init__(
+        self,
+        output_shape: Sequence[int] | None = None,
+        device: torch.device | str = "cpu",
+        dtype: torch.dtype = torch.float32,
+    ) -> None:
+        super().__init__(device, dtype)
+        self.output_shape = output_shape
+
+    def initialize(self, output_shape: Sequence[int] | None = None) -> None:
+        """
+        Initialize the merger by creating tensors for aggregating `values` and `counts`.
+
+        Args:
+            output_shape: the shape of output tensor (B,C,H,W,[D])
+        """
+        # overwrite output_shape if the self.output_shape is set
+        if self.output_shape:
+            output_shape = self.output_shape
+        if output_shape is None:
+            raise ValueError("A valid `output_shape` should be set in the init or initialize method.")
+        self.values = torch.zeros(output_shape, dtype=self.dtype, device=self.device)
+        self.counts = torch.zeros(output_shape, dtype=torch.uint8, device=self.device)
```

Comment:
Is uint8 enough? Is there ever a situation to use this class with more than 255 values to merge?

Reply:
This is essentially the count of overlapping regions, so typycally it should be fine but I can imagine a scenario that someone run infernce with patches of 1000x1000 with 0.999 overlap! I'll change it. Thanks

Reply:
Do you think `torch.int16` (max=32767) would suffice? I want to use the smallest practical dtype as it supposed to be a very large tensor.

Reply:
I made it customizable but default to `torch.int16`.